package/src/package.rs

```diff
@@ -0,0 +1,296 @@
+    pub fn unwrap_library(self) -> Arc<Library> {
+        match self {
+            Self::Executable(_) => panic!("attempted to unwrap 'mast' program as library"),
+            Self::Library(lib) => lib,
+        }
+    }
+
+    /// Get the content digest associated with this artifact
+    pub fn digest(&self) -> Digest {
+        match self {
+            Self::Executable(ref prog) => prog.hash(),
+            Self::Library(ref lib) => *lib.digest(),
+        }
+    }
+
+    /// Get the underlying [MastForest] for this artifact
+    pub fn mast_forest(&self) -> &MastForest {
+        match self {
+            Self::Executable(ref prog) => prog.mast_forest(),
+            Self::Library(ref lib) => lib.mast_forest(),
+        }
+    }
+}
+
+// PACKAGE MANIFEST
+// ================================================================================================
+
+/// The manifest of a package, containing the set of package dependencies (libraries or packages)
+/// and exported procedures and their signatures, if known.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct PackageManifest {
+    /// The set of exports in this package.
+    pub exports: BTreeSet<PackageExport>,
+    /// The libraries (packages) linked against by this package, which must be provided when
+    /// executing the program.
+    pub dependencies: Vec<Dependency>,
+}
+
+/// A procedure exported by a package, along with its digest and
+/// signature(will be added after MASM type attributes are implemented).
+#[derive(Clone, PartialEq, Eq, Ord, PartialOrd)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct PackageExport {
+    /// The fully-qualified name of the procedure exported by this package
+    pub name: QualifiedProcedureName,
+    /// The digest of the procedure exported by this package
+    #[cfg_attr(test, proptest(value = "Digest::default()"))]
+    pub digest: Digest,
+}
+
+impl fmt::Debug for PackageExport {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let Self { name, digest } = self;
+        f.debug_struct("PackageExport")
+            .field("name", &format_args!("{}", name))
+            .field("digest", &format_args!("{}", DisplayHex::new(&digest.as_bytes())))
+            .finish()
+    }
+}
+
+// PACKAGE
+// ================================================================================================
+
+/// A package containing a [Program]/[Library], and a manifest (exports and dependencies).
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+pub struct Package {
+    /// Name of the package
+    pub name: String,
+    /// The MAST artifact ([Program] or [Library]) of the package
+    pub mast: MastArtifact,
+    /// The package manifest, containing the set of exported procedures and their signatures,
+    /// if known.
+    pub manifest: PackageManifest,
+}
```

Comment:
Probably not for this PR, but where would we put custom metadata? Here, I'm thinking about the storage layout metadata we'd need to instantiate account components.

Reply:
I'm thinking, either directly in `Package` or in `PackageManifest`.

Reply:
Let's create an issue for this.

Reply:
Done - https://github.com/0xPolygonMiden/miden-vm/issues/1662